server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java

```diff
@@ -0,0 +1,472 @@
+
+  public SegmentLoader(SegmentStateManager stateManager, DruidCoordinatorRuntimeParams runParams)
+  {
+    this.stateManager = stateManager;
+    this.strategy = runParams.getBalancerStrategy();
+    this.cluster = runParams.getDruidCluster();
+    this.replicantLookup = runParams.getSegmentReplicantLookup();
+    this.maxLoadQueueSize = runParams.getCoordinatorDynamicConfig()
+                                     .getMaxSegmentsInNodeLoadingQueue();
+  }
+
+  public CoordinatorStats getStats()
+  {
+    return stats;
+  }
+
+  /**
+   * Moves the given segment between two servers of the same tier.
+   * <p>
+   * See if we can move balancing here.
+   */
+  public boolean moveSegment(DataSegment segment, ServerHolder fromServer, ServerHolder toServer)
+  {
+    if (!fromServer.getServer().getTier().equals(toServer.getServer().getTier())) {
+      return false;
+    }
+
+    // fromServer must be loading or serving the segment
+    // and toServer must be able to load it
+    final SegmentState stateOnSrc = fromServer.getSegmentState(segment);
+    if ((stateOnSrc != SegmentState.LOADING && stateOnSrc != SegmentState.LOADED)
+        || !canLoadSegment(toServer, segment)) {
+      return false;
+    }
+
+    final boolean cancelSuccess = stateOnSrc == SegmentState.LOADING
+                                  && stateManager.cancelOperation(SegmentState.LOADING, segment, fromServer);
+
+    if (cancelSuccess) {
+      int loadedCountOnTier = replicantLookup
+          .getLoadedReplicants(segment.getId(), toServer.getServer().getTier());
+      stateManager.loadSegment(segment, toServer, loadedCountOnTier < 1);
+    } else {
+      return stateManager.moveSegment(segment, fromServer, toServer);
+    }
+
+    return true;
+  }
+
+  /**
+   * Queues load or drop of replicas of the given segment to achieve the
+   * target replication level in all the tiers.
+   */
+  public void updateReplicas(DataSegment segment, Map<String, Integer> tierToReplicaCount)
+  {
+    // Handle every target tier
+    tierToReplicaCount.forEach((tier, numReplicas) -> {
+      updateReplicasOnTier(segment, tier, tierToReplicaCount.get(tier));
+      stats.addToTieredStat(CoordinatorStats.REQUIRED_CAPACITY, tier, segment.getSize() * numReplicas);
+    });
+
+    // Find the minimum number of segments required for fault tolerance
+    final int totalTargetReplicas = tierToReplicaCount.values().stream()
+                                                      .reduce(0, Integer::sum);
+    final int minLoadedSegments = totalTargetReplicas > 1 ? 2 : 1;
+
+    // Drop segment from unneeded tiers if requirement is met across target tiers
+    int loadedTargetReplicas = 0;
+    final Set<String> targetTiers = tierToReplicaCount.keySet();
+    for (String tier : targetTiers) {
+      loadedTargetReplicas += replicantLookup.getLoadedReplicants(segment.getId(), tier);
+    }
+    if (loadedTargetReplicas < minLoadedSegments) {
+      return;
+    }
```

Comment:
I'm perhaps missing something here, but the logic seems to be saying "if the replication factor is set higher than 1, make sure that there are 2 replicas and no more", which seems... weird.  We should have the target number that the user told us to have?

Reply:
This is the check for queueing up drops in unwanted tiers. If there is even the minimum level of fault tolerance (2 replicas) in the wanted tiers, we can proceed with the drops on unwanted tiers while full replication continues on the wanted tiers.

The assignment to new target tiers is already done by this point.

Reply:
Fixed this to always maintain the currently configured level of replication.